x/marker/keeper/send_restrictions_test.go

```diff
@@ -0,0 +1,409 @@
+			expectedResult: true,
+		},
+		{
+			name:         "should succeed - required with wildcard is contained in attributes",
+			requiredAtts: []string{"*.provenance.io"},
+			attributes: []attrTypes.Attribute{
+				{Name: "kyc.provenance.io"},
+			},
+			expectedResult: true,
+		},
+		{
+			name:         "should succeed - required is contained in attributes",
+			requiredAtts: []string{"kyc.provenance.io"},
+			attributes: []attrTypes.Attribute{
+				{Name: "kyc.provenance.io"},
+			},
+			expectedResult: true,
+		},
+		{
+			name:         "should succeed - multiple attrs and required attrs",
+			requiredAtts: []string{"kyc.provenance.io", "kyc.provenance.com", "kyc.provenance.net"},
+			attributes: []attrTypes.Attribute{
+				{Name: "kyc.provenance.io"},
+				{Name: "kyc.provenance.com"},
+				{Name: "kyc.provenance.net"},
+				{Name: "kyc.provenance.de"},
+			},
+			expectedResult: true,
+		},
+		{
+			name:           "should fail - missing required attr",
+			requiredAtts:   []string{"kyc.provenance.io", "non-kyc.provenance.io"},
+			attributes:     []attrTypes.Attribute{{Name: "kyc.provenance.io"}},
+			expectedResult: false,
+		},
+		{
+			name:           "should fail - missing required attr with wildcard",
+			requiredAtts:   []string{"*.provenance.io", "non-kyc.provenance.io"},
+			attributes:     []attrTypes.Attribute{{Name: "kyc.provenance.io"}},
+			expectedResult: false,
+		},
+	}
+	for _, tc := range testCases {
+		result := keeper.EnsureAllRequiredAttributesExist(tc.requiredAtts, tc.attributes)
+		require.Equal(t, tc.expectedResult, result, fmt.Sprintf("%s", tc.name))
+	}
+}
+
+func TestMatchAttribute(t *testing.T) {
+	testCases := []struct {
+		name           string
+		reqAttr        string
+		attr           string
+		expectedResult bool
+	}{
+		{
+			name:           "should succeed - wildcard on single name",
+			reqAttr:        "*.provenance.io",
+			attr:           "test.provenance.io",
+			expectedResult: true,
+		},
+		{
+			name:           "should succeed - wildcard on multiple names",
+			reqAttr:        "*.provenance.io",
+			attr:           "test.test.test.provenance.io",
+			expectedResult: true,
+		},
+		{
+			name:           "should succeed - literal match",
+			reqAttr:        "test.provenance.io",
+			attr:           "test.provenance.io",
+			expectedResult: true,
+		},
+		{
+			name:           "should fail - wildcard match",
+			reqAttr:        "*.provenance.io",
+			attr:           "test.provenance.com",
+			expectedResult: false,
+		},
+		{
+			name:           "should fail - literal match",
+			reqAttr:        "test.provenance.io",
+			attr:           "test.provenance.com",
+			expectedResult: false,
+		},
+		{
+			name:           "should fail - empty required attr",
+			reqAttr:        "",
+			attr:           "test.provenance.com",
+			expectedResult: false,
+		},
+		{
+			name:           "should fail - empty required attr and attr ",
+			reqAttr:        "",
+			attr:           "",
+			expectedResult: false,
+		},
```

Comment:
Some other test cases that would be good:

```golang
		{
			name:           "should fail - extra ending",
			reqAttr:        "test.provenance.io",
			attr:           "test.provenance.iox",
			expectedResult: false,
		},
		{
			name:           "should fail - wildcard extra ending",
			reqAttr:        "*.provenance.io",
			attr:           "test.provenance.iox",
			expectedResult: false,
		},
		{
			name:           "should fail - wildcard extra beginning",
			reqAttr:        "*.provenance.io",
			attr:           "test.xprovenance.io",
			expectedResult: false,
		},
```

Reply:
Thanks! Your `"should fail - wildcard extra beginning",` found an issue.  Fixed now.